File: frontends/site/src/lib/types/community.ts
export interface Board {
  id: string;
  slug: string;
  name: string;
  description?: string;
  category?: string;
  display_order?: number;
  is_public?: boolean;
  allow_anonymous?: boolean;
  // 파일 업로드 설정
  allow_file_upload?: boolean;
  max_files?: number;
  max_file_size?: number;
  allowed_file_types?: string[];
  allowed_iframe_domains?: string[];
  // 리치 텍스트 에디터 설정
  allow_rich_text?: boolean;
  // 기타 설정
  require_category?: boolean;
  allow_comments?: boolean;
  allow_likes?: boolean;
  // 권한 설정
  write_permission?: string; // guest, member, admin
  list_permission?: string; // guest, member, admin
  read_permission?: string; // guest, member, admin
  reply_permission?: string; // guest, member, admin
  comment_permission?: string; // guest, member, admin
  download_permission?: string; // guest, member, admin
  // 표시 설정
  hide_list?: boolean;
  editor_type?: string; // rich, simple, markdown
  allow_search?: boolean;
  allow_recommend?: boolean;
  allow_disrecommend?: boolean;
  show_author_name?: boolean;
  show_ip?: boolean;
  edit_comment_limit?: number;
  delete_comment_limit?: number;
  use_sns?: boolean;
  use_captcha?: boolean;
  title_length?: number;
  posts_per_page?: number;
  // 포인트 설정
  read_point?: number;
  write_point?: number;
  comment_point?: number;
  download_point?: number;
  created_at?: string;
  updated_at?: string;
}

export interface Category {
  id: string;
  board_id: string;
  name: string;
  description?: string;
  display_order?: number;
  is_active?: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface ThumbnailUrls {
  thumb?: string;   // 목록용 (150x150)
  card?: string;    // 카드용 (300x200)
  large?: string;   // 본문용 (800x600)
}

export interface AttachedFile {
  id: string;
  original_name: string;
  file_path: string;
  file_size: number;
  mime_type: string;
  file_purpose: 'Attachment' | 'Thumbnail' | 'Avatar' | 'EditorImage';
  display_order: number;
}

export interface Post {
  id: string;
  board_id: string;
  category_id?: string;
  user_id: string;
  title: string;
  content: string;
  views?: number;
  likes?: number;
  comments?: number;
  comment_count?: number;
  user_name?: string;
  board_name?: string;
  board_slug?: string;
  category_name?: string;
  is_notice?: boolean;
  status?: string;
  created_at?: string;
  updated_at?: string;
  attached_files?: AttachedFile[];
  thumbnail_urls?: ThumbnailUrls;
}

export interface PostDetail extends Post {
  user_name: string;
  board_name: string;
  board_slug: string;
  category_name?: string;
  comment_count?: number;
  user_id: string;
  board_id: string;
  category_id?: string;
  created_at: string;
  updated_at: string;
  attached_files?: AttachedFile[];
}

export interface Comment {
  id: string;
  post_id: string;
  user_id: string;
  parent_id?: string;
  content: string;
  likes?: number;
  status?: string;
  created_at?: string;
  updated_at?: string;
}

export interface CommentDetail extends Comment {
  user_name: string;
  user_id: string;
  created_at: string;
  updated_at: string;
}

export interface BoardStats {
  board_id: string;
  board_name: string;
  post_count?: number;
  comment_count?: number;
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
  pagination?: any;
}

export interface Category {
  id: string;
  name: string;
  description?: string;
  postCount: number;
}

export interface Tag {
  id: string;
  name: string;
  postCount: number;
}

export type PostSortOption = 'latest' | 'popular' | 'comments';
export interface PostFilter {
  search: string;
  board_id?: string;
  category_id?: string;
  tags: string[];
  sort: 'latest' | 'popular' | 'comments';
  page: number;
  limit: number;
}

export interface PostsResponse {
  posts: Post[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

File: backends/api/src/models/community.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use chrono::{DateTime, Utc};
use uuid::Uuid;

// 게시판 모델
#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Board {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub category: Option<String>,
    pub display_order: Option<i32>,
    pub is_public: Option<bool>,
    pub allow_anonymous: Option<bool>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

// 카테고리 모델
#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Category {
    pub id: Uuid,
    pub board_id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub display_order: Option<i32>,
    pub is_active: Option<bool>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

// 게시글 모델
#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Post {
    pub id: Uuid,
    pub board_id: Uuid,
    pub category_id: Option<Uuid>,
    pub user_id: Uuid,
    pub title: String,
    pub content: String,
    pub views: Option<i32>,
    pub likes: Option<i32>,
    pub is_notice: Option<bool>,
    pub status: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

// 게시글 상세 정보 (사용자 정보 포함)
#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct PostDetail {
    pub id: Uuid,
    pub board_id: Uuid,
    pub category_id: Option<Uuid>,
    pub user_id: Uuid,
    pub title: String,
    pub content: String,
    pub views: Option<i32>,
    pub likes: Option<i32>,
    pub is_notice: Option<bool>,
    pub status: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
    pub user_name: Option<String>,
    pub board_name: Option<String>,
    pub category_name: Option<String>,
    pub comment_count: Option<i64>,
}

// 댓글 모델
#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Comment {
    pub id: Uuid,
    pub post_id: Uuid,
    pub user_id: Uuid,
    pub parent_id: Option<Uuid>,
    pub content: String,
    pub likes: Option<i32>,
    pub status: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

// 댓글 상세 정보 (사용자 정보 포함)
#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct CommentDetail {
    pub id: Uuid,
    pub post_id: Uuid,
    pub user_id: Uuid,
    pub parent_id: Option<Uuid>,
    pub content: String,
    pub likes: Option<i32>,
    pub status: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
    pub user_name: String,
}

// 게시글 생성 요청
#[derive(Debug, Deserialize)]
pub struct CreatePostRequest {
    pub board_id: Uuid,
    pub category_id: Option<Uuid>,
    pub title: String,
    pub content: String,
    pub is_notice: Option<bool>,
}

// 게시글 수정 요청
#[derive(Debug, Deserialize)]
pub struct UpdatePostRequest {
    pub board_id: Option<Uuid>,
    pub category_id: Option<Uuid>,
    pub title: Option<String>,
    pub content: Option<String>,
    pub is_notice: Option<bool>,
}

// 댓글 생성 요청
#[derive(Debug, Deserialize)]
pub struct CreateCommentRequest {
    pub post_id: Uuid,
    pub parent_id: Option<Uuid>,
    pub content: String,
}

// 댓글 수정 요청
#[derive(Debug, Deserialize)]
pub struct UpdateCommentRequest {
    pub content: String,
}

// 게시글 목록 조회 필터
#[derive(Debug, Deserialize)]
pub struct PostFilter {
    pub board_id: Option<Uuid>,
    pub user_id: Option<Uuid>,
    pub status: Option<String>,
    pub search: Option<String>,
    pub sort: Option<String>,
    pub page: Option<i32>,
    pub limit: Option<i32>,
}

// 게시판 목록 응답
#[derive(Debug, Serialize)]
pub struct BoardListResponse {
    pub boards: Vec<Board>,
    pub total: i64,
}

// 게시글 목록 응답
#[derive(Debug, Serialize)]
pub struct PostListResponse {
    pub posts: Vec<PostDetail>,
    pub total: i64,
    pub page: i64,
    pub limit: i64,
    pub total_pages: i64,
}

// 댓글 목록 응답
#[derive(Debug, Serialize)]
pub struct CommentListResponse {
    pub comments: Vec<CommentDetail>,
    pub total: i64,
}

// 최근 게시글 응답
#[derive(Debug, Serialize)]
pub struct RecentPostsResponse {
    pub posts: Vec<PostDetail>,
}

// 게시판 통계
#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct BoardStats {
    pub board_id: Uuid,
    pub board_name: String,
    pub post_count: Option<i64>,
    pub comment_count: Option<i64>,
}

// 게시글 조회 쿼리
#[derive(Debug, Serialize, Deserialize)]
pub struct PostQuery {
    pub search: Option<String>,
    pub board_id: Option<Uuid>,
    pub category_id: Option<Uuid>,
    pub sort: Option<String>,
    pub page: Option<i64>,
    pub limit: Option<i64>,
}

// 게시글 간단 요약 (테스트용)
#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct PostSummary {
    pub id: Uuid,
    pub title: String,
    pub board_id: Option<Uuid>,
    pub user_name: Option<String>,
    pub board_name: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
} 